Sources/ConnectionPoolModule/PoolStateMachine+ConnectionGroup.swift

```diff
@@ -0,0 +1,634 @@
+            self.timersToCancel = timersToCancel
+            self.wasIdle = wasIdle
+            self.use = use
+        }
+    }
+
+    @usableFromInline
+    struct ConnectionGroup {
+        @usableFromInline
+        struct Stats: Hashable {
+            @usableFromInline var connecting: UInt16 = 0
+            @usableFromInline var backingOff: UInt16 = 0
+            @usableFromInline var idle: UInt16 = 0
+            @usableFromInline var leased: UInt16 = 0
+            @usableFromInline var runningKeepAlive: UInt16 = 0
+            @usableFromInline var closing: UInt16 = 0
+
+            @usableFromInline var availableStreams: UInt16 = 0
+            @usableFromInline var leasedStreams: UInt16 = 0
+
+            @usableFromInline var soonAvailable: UInt16 {
+                self.connecting + self.backingOff + self.runningKeepAlive
+            }
+
+            @usableFromInline var active: UInt16 {
+                self.idle + self.leased + self.connecting + self.backingOff
+            }
+        }
+
+        /// The minimum number of connections
+        @usableFromInline
+        let minimumConcurrentConnections: Int
+
+        /// The maximum number of preserved connections
+        @usableFromInline
+        let maximumConcurrentConnectionSoftLimit: Int
+
+        /// The absolute maximum number of connections
+        @usableFromInline
+        let maximumConcurrentConnectionHardLimit: Int
+
+        @usableFromInline
+        let keepAlive: Bool
+
+        @usableFromInline
+        let keepAliveReducesAvailableStreams: Bool
+
+        /// A connectionID generator.
+        @usableFromInline
+        let generator: ConnectionIDGenerator
+
+        /// The connections states
+        @usableFromInline
+        private(set) var connections: [ConnectionState]
+
+        @usableFromInline
+        private(set) var stats = Stats()
+
+        @inlinable
+        init(
+            generator: ConnectionIDGenerator,
+            minimumConcurrentConnections: Int,
+            maximumConcurrentConnectionSoftLimit: Int,
+            maximumConcurrentConnectionHardLimit: Int,
+            keepAlive: Bool,
+            keepAliveReducesAvailableStreams: Bool
+        ) {
+            self.generator = generator
+            self.connections = []
+            self.minimumConcurrentConnections = minimumConcurrentConnections
+            self.maximumConcurrentConnectionSoftLimit = maximumConcurrentConnectionSoftLimit
+            self.maximumConcurrentConnectionHardLimit = maximumConcurrentConnectionHardLimit
+            self.keepAlive = keepAlive
+            self.keepAliveReducesAvailableStreams = keepAliveReducesAvailableStreams
+        }
+
+        var isEmpty: Bool {
+            self.connections.isEmpty
+        }
+
+        @usableFromInline
+        var canGrow: Bool {
+            self.stats.active < self.maximumConcurrentConnectionHardLimit
+        }
+
+        @usableFromInline
+        var soonAvailable: UInt16 {
```

Comment:
This name sounds more like something boolean than a counter. Is there something more specific to call this?